kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/stores/AsyncKeyValueStore.java

```diff
@@ -82,10 +84,12 @@ public void executeDelayedWrite(final DelayedWrite<KS, VS> delayedWrite) {
   @Override
   public void put(KS key, VS value) {
     final Thread currentThread = Thread.currentThread();
-
-    if (currentThread instanceof AsyncThread) {
-      final AsyncEvent currentAsyncEvent =
-          ((AsyncThread) currentThread).context(partition).currentAsyncEvent();
+    final MergedProcessorContext<?, ?> rawThreadCtx = userProcessorContext.delegate();
+    if (rawThreadCtx instanceof AsyncThreadProcessorContext) {
+      // todo: yuck
+      final AsyncThreadProcessorContext<?, ?> threadCtx = (AsyncThreadProcessorContext<?, ?>)
+          userProcessorContext.delegate();
```

Comment:
I really agree with the "yuck" here -- I do worry that I'm a bit biased in favor of the way I wrote it originally just because that's now what feels familiar to me, but I really don't like the idea of exposing/using the user context router (ie the AsyncUserProcessorContext) outside of the user's processor. I think it's confusing enough that the context we pass in to the user is going to return a different actual underlying context depending on which thread is currently executing, and I'd prefer to keep that isolated to only what we pass in to the user and not use the "thread-local routing" mechanism within the async processing framework

I'll acknowledge that what I was doing before was exactly the same in practice: grabbing the context for the specific thread that's executing. I just think it's a lot more clear which specific context we're accessing here when you do something like ((AsyncThread) currentThread).context() vs asking for the "delegate context" from some other context which is really obtuse until you already understand exactly what everything is doing and how all the different contexts interact.

I'm also maybe a bit biased because from lurking on various AK tickets I know the use of ThreadLocal has caused some major bugs and general confusion in core Kafka. I'm not against the use of ThreadLocal itself, but I do think it's just somewhat of a non-obvious behavior and the whole which-context-is-which thing is already confusing enough given the many different contexts and types of contexts we're dealing with in the async framework

Reply:
I agree with your thoughts here - especially the bit about breaking the context layering encapsulation. I tried to address this in the latest patch by adding a new interface called `DelayedAsyncStoreWriter`, which a caller can use to do a delayed write to the state store. This is what gets used by the async store. It's implemented by the user context, which does the delayed write to the current event if its called on the async thread.